third_party/tonic/tests/ffi_native_unittest.cc

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/testing/dart_isolate_runner.h"
+#include "flutter/testing/fixture_test.h"
+
+#include "tonic/dart_args.h"
+#include "tonic/dart_wrappable.h"
+
+namespace flutter {
+namespace testing {
+
+class FfiNativeTest : public FixtureTest {
+ public:
+  FfiNativeTest()
+      : settings_(CreateSettingsForFixture()),
+        vm_(DartVMRef::Create(settings_)),
+        thread_(CreateNewThread()),
+        task_runners_(GetCurrentTestName(),
+                      thread_,
+                      thread_,
+                      thread_,
+                      thread_) {}
+
+  ~FfiNativeTest() = default;
+
+  [[nodiscard]] bool RunWithEntrypoint(const std::string& entrypoint) {
+    if (running_isolate_) {
+      return false;
+    }
+    auto isolate =
+        RunDartCodeInIsolate(vm_, settings_, task_runners_, entrypoint, {},
+                             GetDefaultKernelFilePath());
+    if (!isolate || isolate->get()->GetPhase() != DartIsolate::Phase::Running) {
+      return false;
+    }
+
+    running_isolate_ = std::move(isolate);
+    return true;
+  }
+
+ protected:
+  Settings settings_;
+  DartVMRef vm_;
+  std::unique_ptr<AutoIsolateShutdown> running_isolate_;
+  fml::RefPtr<fml::TaskRunner> thread_;
+  TaskRunners task_runners_;
+  FML_DISALLOW_COPY_AND_ASSIGN(FfiNativeTest);
+};
+
+fml::AutoResetWaitableEvent event;
+
+void Nop() {
+  event.Signal();
+}
+
+// Make a simple call through the Tonic FFI bindings.
+TEST_F(FfiNativeTest, FfiBindingCallNop) {
+  AddFfiNativeCallback(
+      "Nop", reinterpret_cast<void*>(
+                 tonic::FfiDispatcher<void, decltype(&Nop), &Nop>::Call));
+
+  ASSERT_TRUE(RunWithEntrypoint("nop"));
+  event.Wait();
+
+  running_isolate_->Shutdown();
+}
+
+Dart_Handle Echo(Dart_Handle str) {
+  return str;
+}
+
+// Make call with handles through the Tonic FFI bindings.
+TEST_F(FfiNativeTest, FfiBindingCallEcho) {
+  AddFfiNativeCallback(
+      "Echo", reinterpret_cast<void*>(
+                  tonic::FfiDispatcher<void, decltype(&Echo), &Echo>::Call));
+
+  AddNativeCallback(
+      "SignalDone",
+      CREATE_NATIVE_ENTRY([&](Dart_NativeArguments args) { event.Signal(); }));
+
+  ASSERT_TRUE(RunWithEntrypoint("callEcho"));
+  event.Wait();
+
+  running_isolate_->Shutdown();
+}
+
+void MyTestFunction(intptr_t a, float b) {}
+
+// Serialise a static function without handles.
```

Comment:
Can we also make these serialize tests do an FFI call into the functions we serialize for?